File: 6.Software-Design-Patterns/example/Creational/Singleton/Singleton.py
class SingletonMeta(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            instance = super().__call__(*args, **kwargs)
            cls._instances[cls] = instance
        return cls._instances[cls]


class Singleton(metaclass=SingletonMeta):
    def some_business_logic(self):


if __name__ == "__main__":
    s1 = Singleton()
    s2 = Singleton()

    if id(s1) == id(s2):
        print("Singleton works, both variables contain the same instance.")
    else:
        print("Singleton failed, variables contain different instances.")

File: 2.Intermediate Level/6.Software-Design-Patterns/example/Creational/Singleton/Singleton.js
class SingletonJs {
  static instance = null;

  constructor() {
    if (SingletonJs.instance) {
      throw new Error(
        "Singleton cannot be instantiated directly. Use getInstance() instead."
      );
    }
    SingletonJs.instance = this;
  }

  static getInstance() {
    if (!SingletonJs.instance) {
      SingletonJs.instance = new SingletonJs();
    }
    return SingletonJs.instance;
  }
}

function clientJs() {
  const example1 = SingletonJs.getInstance();
  const example2 = SingletonJs.getInstance();
  if (example1 === example2) {
    console.log("Singleton works, both variables contain the same instance.");
  } else {
    console.log("Singleton failed, variables contain different instances.");
  }
}

clientJs();
